File: 03-apps/bergastream/backend/app/models/track.py
from datetime import datetime
from sqlalchemy import String, Boolean, Integer, BigInteger, DateTime, func
from sqlalchemy.orm import Mapped, mapped_column, relationship
from app.database import Base


class Track(Base):
    __tablename__ = "tracks"

    # ID format: "deezer_123456789" | "youtube_abc123" | "spotify_abc123"
    id: Mapped[str] = mapped_column(String(100), primary_key=True)
    title: Mapped[str] = mapped_column(String, nullable=False)
    artist: Mapped[str] = mapped_column(String, nullable=False)
    album: Mapped[str | None] = mapped_column(String, nullable=True)
    album_id: Mapped[str | None] = mapped_column(String, nullable=True)
    artist_id: Mapped[str | None] = mapped_column(String, nullable=True)
    duration_ms: Mapped[int | None] = mapped_column(Integer, nullable=True)
    year: Mapped[int | None] = mapped_column(Integer, nullable=True)
    cover_url: Mapped[str | None] = mapped_column(String, nullable=True)
    source: Mapped[str] = mapped_column(String(20), nullable=False)  # 'deezer' | 'youtube' | 'spotify'
    source_id: Mapped[str | None] = mapped_column(String, nullable=True)
    file_path: Mapped[str | None] = mapped_column(String, nullable=True)   # /data/music/permanent/
    cache_path: Mapped[str | None] = mapped_column(String, nullable=True)  # /data/music/cache/
    is_permanent: Mapped[bool] = mapped_column(Boolean, default=False)
    cache_expires_at: Mapped[datetime | None] = mapped_column(DateTime(timezone=True), nullable=True)
    file_size_bytes: Mapped[int | None] = mapped_column(BigInteger, nullable=True)
    audio_quality: Mapped[str | None] = mapped_column(String(10), nullable=True)  # 'flac' | 'mp3_320' | 'mp3_128'
    created_at: Mapped[datetime] = mapped_column(DateTime(timezone=True), server_default=func.now())
    last_accessed_at: Mapped[datetime] = mapped_column(DateTime(timezone=True), server_default=func.now(), onupdate=func.now())

    playlist_tracks: Mapped[list["PlaylistTrack"]] = relationship("PlaylistTrack", back_populates="track")
    liked_by: Mapped[list["LikedSong"]] = relationship("LikedSong", back_populates="track")
    play_history: Mapped[list["PlayHistory"]] = relationship("PlayHistory", back_populates="track")
    offline_tracks: Mapped[list["OfflineTrack"]] = relationship("OfflineTrack", back_populates="track")

    @property
    def current_file_path(self) -> str | None:
        """Returns the available file path (permanent > cache)."""
        return self.file_path or self.cache_path

File: 03-apps/bergastream/frontend/lib/models/track.dart
class Track {
  final String id;
  final String title;
  final String artist;
  final String? album;
  final String? albumId;
  final String? artistId;
  final int? durationMs;
  final int? year;
  final String? coverUrl;
  final String source;
  final String? sourceId;
  final bool isPermanent;
  final String? audioQuality;

  const Track({
    required this.id,
    required this.title,
    required this.artist,
    this.album,
    this.albumId,
    this.artistId,
    this.durationMs,
    this.year,
    this.coverUrl,
    required this.source,
    this.sourceId,
    this.isPermanent = false,
    this.audioQuality,
  });

  factory Track.fromJson(Map<String, dynamic> json) => Track(
    id: json['id'] as String,
    title: json['title'] as String,
    artist: json['artist'] as String,
    album: json['album'] as String?,
    albumId: json['album_id'] as String?,
    artistId: json['artist_id'] as String?,
    durationMs: json['duration_ms'] as int?,
    year: json['year'] as int?,
    coverUrl: json['cover_url'] as String?,
    source: json['source'] as String,
    sourceId: json['source_id'] as String?,
    isPermanent: json['is_permanent'] as bool? ?? false,
    audioQuality: json['audio_quality'] as String?,
  );

  Map<String, dynamic> toJson() => {
    'id': id,
    'title': title,
    'artist': artist,
    'album': album,
    'album_id': albumId,
    'artist_id': artistId,
    'duration_ms': durationMs,
    'year': year,
    'cover_url': coverUrl,
    'source': source,
    'source_id': sourceId,
    'is_permanent': isPermanent,
    'audio_quality': audioQuality,
  };

  String get durationFormatted {
    if (durationMs == null) return '--:--';
    final s = durationMs! ~/ 1000;
    final m = s ~/ 60;
    final sec = s % 60;
    return '$m:${sec.toString().padLeft(2, '0')}';
  }
}
